src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+///
+///  - client arguments: The list of arguments that will be used to run the jvm program under nailgun.
+///    These arguments can be thought of as "passthrough args" that are sent to the jvm via the nailgun client.
+///    These arguments include everything starting from the main class.
+///
+/// We assume that:
+///  - Every args list has a main class.
+///  - There is exactly one argument that doesn't begin with a `-` in the command line before the main class,
+///    and it's the value of the classpath (i.e. `-cp scala-library.jar`).
+///
+/// We think these assumptions are valid as per: https://github.com/pantsbuild/pants/issues/8387
+///
+fn split_args(args: &Vec<String>) -> ParsedArgLists {
+    let mut nailgun_args = vec![];
+    let mut client_args = vec![];
+    let mut have_seen_classpath = false;
+    let mut have_processed_classpath = false;
+    let mut have_seen_main_class = false;
+    for arg in args {
+        if have_seen_main_class {
+            // Arguments to pass to the client request
+            client_args.push(arg.clone());
+        } else if (arg == "-cp" || arg == "-classpath") && !have_seen_classpath {
+            // Process the -cp flag itself
+            have_seen_classpath = true;
+            nailgun_args.push(arg.clone());
+        } else if have_seen_classpath && !have_processed_classpath {
+            // Process the classpath string
+            nailgun_args.push(arg.clone());
+            have_processed_classpath = true;
+        } else if have_processed_classpath && !arg.starts_with("-") {
+            // Process the main class:
+            // I have already seen the value of the -cp classpath, and this is not a flag.
+            client_args.push(arg.clone());
+            have_seen_main_class = true;
+        } else {
+            // Process the rest (jvm options to pass to nailgun)
+            nailgun_args.push(arg.clone());
+        }
+    }
+    ParsedArgLists {
+        nailgun_args: nailgun_args,
+        client_args: client_args
+    }
+}
+
+// TODO(8481) The input_files arg should be calculated from deeply fingerprinting the classpath.
+fn get_nailgun_request(args: Vec<String>, _input_files: Digest, jdk: Option<PathBuf>) -> ExecuteProcessRequest {
+    let mut full_args = args;
+    full_args.push(NAILGUN_MAIN_CLASS.to_string());
+    full_args.extend(ARGS_TO_START_NAILGUN.iter().map(|a| a.to_string()));
+
+    ExecuteProcessRequest {
+        argv: full_args,
+        env: BTreeMap::new(),
+        input_files: hashing::EMPTY_DIGEST,
+        output_files: BTreeSet::new(),
+        output_directories: BTreeSet::new(),
+        timeout: Duration::new(1000, 0),
+        description: String::from("ExecuteProcessRequest to start a nailgun"),
+        unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule: hashing::EMPTY_DIGEST,
+        jdk_home: jdk,
+        target_platform: Platform::Darwin,
```

Comment:
This should probably be `Platform::current_platform()`

Reply:
I think this should take the platform from the client request (updated in a local commit that I haven't pushed yet)